### PR TITLE
fix: improve inferred json encoding for csv output

### PR DESCRIPTION
**Pull request by iand**
Nil JSON values should be encoded as lower-case `null` in the CSV output, but other nil values should remain uppercase

Slices and other non-scalar types are automatically inferred to be stored as JSON by go-pg so we can use that to automatically encode. However `string` types that have the `type:jsonb` annotation in our models refer to values that have already been encoded as JSON, so the CSV output needs to avoid encoding them again.

### CODE REVIEW

File: storage/csv.go [hunk 1]
```diff
@@ -196,6 +194,24 @@ func (c *CSVBatch) PersistModel(ctx context.Context, m interface{}) error {
 				continue
 			}
 
+			var encodeAsJSON bool
+
+			// Strings marked as json type are assumed to already be encoded
+			if fk != reflect.String && (t.types[i] == "json" || t.types[i] == "jsonb") {
+				encodeAsJSON = true
```
Comment (frrist) [inline storage/csv.go:201]: Should this be false if its already encoded?

File: storage/csv.go [hunk 2]
```diff
@@ -196,6 +194,24 @@ func (c *CSVBatch) PersistModel(ctx context.Context, m interface{}) error {
...
+			} else if fk == reflect.Interface {
+				encodeAsJSON = true
+			}
```
Comment (frrist) [inline storage/csv.go:204]: This means slices of strings/bytes/structs will be encoded as json?

**Review by frrist — APPROVED** [review]

File: storage/csv.go (hunk 1 above)
Comment (iand) [reply storage/csv.go:201]: No, this is checking if it has a json datatype but is not a string

Comment (iand) [reply storage/csv.go:201]: Comment should be clearer

File: storage/csv.go (hunk 2 above)
Comment (iand) [reply storage/csv.go:204]: Slices of strings encode as json. For example `MultisigTransaction.Approved` is a `[]string` with no column type tag. go-pg assigns that a json type by default, which matches our schema. Same with `MinerInfo.ControlAddresses` and `MinerInfo.MultiAddresses`

**Pull request merged**